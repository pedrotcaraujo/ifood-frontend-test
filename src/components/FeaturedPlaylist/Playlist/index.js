import React, { Component } from 'react';

class Playlist extends Component {
    render() {
        return (
            <ul>
            {this.props.data.items.map(item => (
                <li key={item.id}>
                    <figure>
                        <img src={item.images[0].url} alt={item.name}/>
                        <figcaption>{item.name}</figcaption>
                    </figure>
                </li>                    
            ))}
            </ul>
        )
    }
}

export default Playlist;