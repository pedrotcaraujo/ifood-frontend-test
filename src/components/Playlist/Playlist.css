.Playlist {
    width: 100%;
}

.Playlist-items {
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    margin: 0;
}

.Playlist-item {
    list-style: none;
}

.Playlist-message {
    font-size: 2em;
    text-shadow: 1px 1px 5px #000;
}

.Playlist-image {
    box-shadow: 1px 1px 50px #000;
}

.Playlist-figcaption {
    margin-top: 10px;
}