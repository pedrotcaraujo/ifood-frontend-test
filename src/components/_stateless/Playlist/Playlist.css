.Playlist {
    margin: 0;
}

.Playlist-play {
    position: relative;
    cursor: pointer;
    display: flex;
    align-items: center;
    justify-content: center;
}

.Playlist-play:hover::after {
    content: ' ';
    position: absolute;
    z-index: 1;
    background: #000;
    height: 101%;
    width: 101%;
    opacity: .8;
    top: -.5px;
    left: -0.5px;
}

.Playlist-play:hover .Playlist-button {
    display: block;
}

.Playlist-image {
    box-shadow: 1px 1px 50px #000;
}

.Playlist-button {
    display: none;
    position: absolute;
    z-index: 2;
    border: 0;
    background: transparent;
    box-sizing: border-box;
    width: 0;
    height: 74px;
  
    border-color: transparent transparent transparent #fff;
    transition: 100ms all ease;
    cursor: pointer;
  
    border-style: solid;
    border-width: 37px 0 37px 60px;
}

.Playlist-figcaption {
    margin-top: 10px;
}