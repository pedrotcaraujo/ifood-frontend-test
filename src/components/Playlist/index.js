import React, { Component } from 'react';
import axios from 'axios';
import './Playlist.css';

import FilterStore from '../../stores/FilterStore';

import Loader from '../Loader';

const URL = 'https://api.spotify.com/v1/browse/featured-playlists';
const PULLING_TIMEOUT = 30000;

class Playlist extends Component {
    constructor() {
        super()
        this.state = {
                loaded: false,
                search: '',
                message: null, 
                playlists: { 
                    items: [] 
                }
            }
        this.timeouts = [];
    }
    componentDidMount() {
        FilterStore.addListener(this.onFilterChange)
        if (this.props.token) {
            window.requestAnimationFrame(() => this.update(this));
        }
    }

    componentWillUnmount() {
        FilterStore.removeListener(this.onFilterChange)
        window.clearTimeout(this.timeout);
    }

    update(context, params = {}) {
        this.setState({ loaded: false })
        axios.get(URL, {
            params,
            headers: {
              'Authorization': 'Bearer ' + context.props.token
            }
        })
        .then(({ data }) => {
            context.setState({ loaded: true, ...data });
            context.pulling(PULLING_TIMEOUT, () => context.update(context, params));
        })
        .catch(this.props.onError);
    }

    pulling(timeout, callback) {
        this.timeouts.push(window.setTimeout(() => window.requestAnimationFrame(callback), timeout));
    }

    onFilterChange = () => {
        const { current, data } = FilterStore.getState();
        const { search, ...params } = data;

        if (current.name === 'search') {
            this.setState(Object.assign({...this.state}, { search: current.value }))
            return;
         }

        if (this.timeouts.length > 0) {
            this.timeouts.forEach(timeout => {
                window.clearTimeout(timeout)
            })
        }

        if (this.state.loaded) {
            this.update(this, params);
        }
    }

    fuzzySearch = ({ name }) => {
        const { search } = this.state;        
        if (!search) { return true };
        const re = new RegExp(search, 'i');
        return name.match(re)
    }

    render() {
        return (
            <Loader loaded={this.state.loaded}>
                <div className="Playlist">
                    <h2 className="Playlist-message">{this.state.message}</h2>
                    <ul className="Playlist-items">
                    {this.state.playlists.items.filter(this.fuzzySearch).map(item => (
                        <li className="Playlist-item" key={item.id}>
                            <figure>
                                <img className="Playlist-image" src={item.images[0].url} alt={item.name}/>
                                <figcaption className="Playlist-figcaption">{item.name}</figcaption>
                            </figure>
                        </li>                    
                    ))}
                    </ul>
                </div>
            </Loader>
        )
    }
}

export default Playlist;