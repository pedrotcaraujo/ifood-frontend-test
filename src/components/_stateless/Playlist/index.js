import React from 'react';
import PropTypes from 'prop-types';
import './Playlist.css';

const Playlist = ({image, name}) => (
    <figure className="Playlist">
        <div className="Playlist-play">
            <button className="Playlist-button"/>
            <img className="Playlist-image" src={image} alt={name}/>
        </div>
        <figcaption className="Playlist-figcaption">{name}</figcaption>
    </figure>
);

Playlist.propTypes = {
    image: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired
};

export default Playlist;