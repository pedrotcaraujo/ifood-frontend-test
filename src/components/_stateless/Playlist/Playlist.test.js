import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Playlist from './index';

it('renders correctly', () => {
  const tree = renderer
    .create(<Playlist image="asdas" name="asd" />)
    .toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Playlist image="asdas" name="asd" />, div);
  ReactDOM.unmountComponentAtNode(div);
});
