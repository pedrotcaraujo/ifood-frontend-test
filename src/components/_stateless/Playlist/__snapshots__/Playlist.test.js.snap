// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<figure
  className="Playlist"
>
  <div
    className="Playlist-play"
  >
    <button
      className="Playlist-button"
    />
    <img
      alt="asd"
      className="Playlist-image"
      src="asdas"
    />
  </div>
  <figcaption
    className="Playlist-figcaption"
  >
    asd
  </figcaption>
</figure>
`;
